File: Data Structures & Algorithms/longest-consecutive-sequence/submission-2.py
class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        d = {}
        res = 0
        for num in nums:
            if num not in d:
                d[num] = d.get(num - 1, 0) + d.get(num + 1, 0) + 1
            d[num - d.get(num - 1, 0)] = d.get(num)
            d[num + d.get(num + 1, 0)] = d.get(num)
            res = max(res, d.get(num))
        return res

File: Data Structures & Algorithms/longest-consecutive-sequence/submission-2.java
class Solution {
    public int longestConsecutive(int[] nums) {
        HashMap<Integer, Integer> seq = new HashMap<>();

        int res = 0;
        for (int num: nums){
            if (seq.containsKey(num)) continue;
            seq.put(num, seq.getOrDefault(num - 1, 0) + seq.getOrDefault(num + 1, 0) + 1);
            // update left boundary;
            seq.put(num - seq.getOrDefault(num - 1, 0), seq.get(num));
            // update right boundary;
            seq.put(num + seq.getOrDefault(num + 1, 0), seq.get(num));
            res = Math.max(res, seq.get(num));
        }
        return res;
    }
}
